modules/util.py

```diff
@@ -124,39 +127,78 @@ class PomUtils:
     def __init__(self, driver: Firefox):
         self.driver = driver
 
-    def get_shadow_content(self, element: WebElement) -> list[WebElement]:
+    def get_shadow_content(
+        self, element: WebElement
+    ) -> list[Union[WebElement, ShadowRoot]]:
         """
         Given a WebElement, return the shadow DOM root or roots attached to it. Returns a list.
         """
-        try:
-            shadow_root = element.shadow_root
-            return [shadow_root]
-        except InvalidArgumentException:
+        logging.info(f"Getting shadow nodes from root {element}")
+
+        def shadow_from_script():
             shadow_children = self.driver.execute_script(
                 "return arguments[0].shadowRoot.children", element
             )
             if len(shadow_children) and any(shadow_children):
-                return [s for s in shadow_children if s is not None]
+                logging.info("Returning script-returned shadow elements")
+                shadow_elements = [s for s in shadow_children if s is not None]
+                logging.info(shadow_elements)
+                return shadow_elements
+
+        try:
+            logging.info("Getting shadow content...")
+            shadow_root = element.shadow_root
+            shadow_content = [shadow_root]
+            if not shadow_content:
+                logging.info("Selenium shadow nav returned no elements in Shadow DOM")
+                return shadow_from_script()
+            return shadow_content
+        except InvalidArgumentException:
+            logging.info("Selenium shadow nav failed.")
+            return shadow_from_script()
         return []
 
+    def css_selector_matches_element(
+        self, element: Union[WebElement, ShadowRoot], selector: list
+    ) -> bool:
+        if type(element) == ShadowRoot:
+            return False
```

Comment:
Could do 
```suggestion
        if element is ShadowRoot:
            return False
```

Reply:
I just committed this, and then it somehow broke a test? That is an odd fact, but a true one. Looks like Selenium didn't properly implement `id` for that type.